### PR TITLE
Migrate to zod

### PR DESCRIPTION
**Pull request by FuzzB0t**
WIP - Ref #270 

Migrated all instances of AJV validation to zod, this PR is still a work in progress as there is more testing and refining to be done.

Here's what's done so far:
- [x] Changed all JTD schemas into zod schemas
- [x] Removed AJV library from repo
- [x] Successfully ran test suite + manual validation testing
- [x] Added docgen for schema types
- [x] Added cannonfile spec view to website

An example of manual validation testing is the following:

Validation Error on wrong required keyname "artifacts":
image

Validation error on invalid required key type:
image

Successful build after fixing all errors:
image

### CODE REVIEW

File: packages/builder/src/actions.ts [hunk 1]
```diff
@@ -46,29 +45,50 @@ export const ActionKinds: { [label: string]: CannonAction } = {};
...
+      invalid_type_error: 'version must be a string',
+    }),
+  })
+  .merge(
+    z
+      .object({
+        description: z.string(),
+        keywords: z.array(
+          z.string({
+            invalid_type_error: 'keywords must be strings',
```
Comment (dbeal-eth) [inline packages/builder/src/actions.ts:66]: Keywords must be an array of strings

File: packages/builder/src/steps/invoke.ts [hunk 2]
```diff
@@ -107,7 +151,7 @@ async function runTxn(
     );
   }
 
-  if (config.fromCall && !contract.functions[config.fromCall.func]) {
+  if (config.fromCall && config.fromCall.func && !contract.functions[config.fromCall.func]) {
```
Comment (dbeal-eth) [inline packages/builder/src/steps/invoke.ts:154]: Did you run into an issue with this? Seems like unrelated change

File: packages/builder/src/types.zod.ts [hunk 3]
```diff
@@ -0,0 +1,280 @@
+import { ethers } from 'ethers';
```
Comment (dbeal-eth) [inline packages/builder/src/types.zod.ts:1]: This is really cool, having all the internal storage types in zod as well. Does this mean we will have a way to validate ipfs artifacts just to make sure? Can we use it as a way to verify that updates to cannon schema are non-breaking progmatically

File: packages/builder/src/util.ts [hunk 4]
```diff
@@ -4,21 +4,38 @@ import { Buffer } from 'buffer';
...
-export const ChainDefinitionScriptSchema = {
-  properties: {
-    exec: { type: 'string' },
-  },
-  optionalProperties: {
-    args: { elements: { type: 'string' } },
-    env: { elements: { type: 'string' } },
-  },
-} as const;
+// FIXME: Why not move this to keepers.ts since this isnt shared by any other file?
```
Comment (dbeal-eth) [inline packages/builder/src/util.ts:15]: This used to be used be used for both keepers and `run` step (which was moved into cli obviously)

So I would agree it makes sense to make this change

**Review by dbeal-eth — COMMENTED** [review]
Looking really good! Not a fine tooth comb reciew but would like to see much better error lofting format (than dumping a Json object). Do you have any ideas for that?

File: packages/builder/src/actions.ts (hunk 1 above)
Comment (FuzzB0t) [reply packages/builder/src/actions.ts:66]: Yeah it is an array of strings, I'll change the error message to better resemble that but this error pops up if any of the keywords inside the array aren't strings.

File: packages/builder/src/steps/invoke.ts (hunk 2 above)
Comment (FuzzB0t) [reply packages/builder/src/steps/invoke.ts:154]: I did, It was potentially coming back undefined since the fromCall parameter is optional, zod handles optional parameters as a union of ```type | undefined```

File: packages/builder/src/types.zod.ts (hunk 3 above)
Comment (FuzzB0t) [reply packages/builder/src/types.zod.ts:1]: Yes! That's the idea, it's not fully implemented yet but the idea is that we can use zod as our source of truth which provides us inherent validation for our core types while being able to generate types directly from our zod schemas.

**Comment by FuzzB0t** [comment]
> Looking really good! Not a fine tooth comb reciew but would like to see much better error lofting format (than dumping a Json object). Do you have any ideas for that?

Yes, so right now im using zod's ```.parse``` method which throws a ZodError object (the JSON dump). I can use ```safeParse``` to silently throw an error and catch these errors and create custom error handlers like in [this part](https://zod.dev/?id=error-formatting) of the zod documentation.

**Comment by FuzzB0t** [comment]
Remaining tasks list for progress reference:

- [ ] Improve error logging
- [ ] Add docgen from zod types (using something like [this](https://github.com/samchungy/zod-openapi))
- [ ] Look into adding type inference from zod types to ts (This is to keep zod as our source of truth and be able to validate results or inputs from things like IPFS or JSON-RPC calls)
- [ ] Add guidelines to builder README for adding new input types

**Comment by FuzzB0t** [comment]
Fixing errors, they now look like this:

image

This format is easily customizable and can be changed to match another kind of pattern.

Also added a way to autogenerate markdown docs from the ts inferred zod schemas.

**Comment by FuzzB0t** [comment]
Added html docgen  with Typedoc:

image

Example Step Config Documentation (HTML):
image

Also added Markdown Docgen (for github):
image

File: packages/builder/src/schemas.zod.ts
Comment (FuzzB0t) [inline packages/builder/src/schemas.zod.ts:1]: Contains all of the step config schemas that were validated with AJV

File: packages/builder/src/schemas.zod.ts [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+  );
+
+/// ================================ TS TYPE DEFINITIONS ================================ \\\
+
+// General validation schema.
+// Anytime an attribute is added to any of the zod schemas above it must be added to this
+
+// All properties are made optional since all schemas that
+// are validated against this are a partial object of this schema.
+export type ConfigValidationSchema = z.ZodObject<{
```
Comment (FuzzB0t) [inline packages/builder/src/schemas.zod.ts:174]: General validation schema to use to validate step configs against

File: packages/website/content/docs-technical.md [hunk 6]
```diff
@@ -68,7 +68,6 @@ Publish a Cannon package to the registry for all networks where this package has
 **Options**
 
 - `--chain-id` - The network which the registry to deploy to is on. Default: 1 (mainnet)
-- `--provider-url` - RPC endpoint to use for publishing. If not specified, uses a local wallet provider (such as Frame)
```
Comment (FuzzB0t) [inline packages/website/content/docs-technical.md:71]: This piece in the docs was incorrect, after looking through the publish.ts file there was no reference to use of the "provider-url" param

**Comment by FuzzB0t** [comment]
Successfully built the synthetix-sandbox package:

```
n00b@DESKTOP-C38BT9L:~/work/cannon-packages/synthetix-sandbox$ cannon-dev build 
Building the foundry project using forge build...
forge build succeeded
Checking IPFS for package synthetix-sandbox:3.1.2...
Existing package found.

Using package...
Name: synthetix-sandbox
Version: 3.1.2
Preset: main (default)
Source code will be included in the package

Building the chain (ID 13370 via frame) into the state defined in cannonfile.toml...
plugins: cannon-plugin-router detected
exec: provision.synthetix
Checking IPFS for package synthetix:latest...
  exec: contract.AccountModule
  exec: contract.AssociateDebtModule
  exec: contract.CcipReceiverModule
  exec: contract.CollateralConfigurationModule
  exec: contract.CollateralModule
  exec: contract.CrossChainUSDModule
  exec: contract.IssueUSDModule
  exec: contract.LiquidationModule
  exec: contract.MarketCollateralModule
  exec: contract.MarketManagerModule
  exec: contract.MulticallModule
  exec: contract.PoolConfigurationModule
  exec: contract.PoolModule
  exec: contract.RewardsManagerModule
  exec: contract.USDTokenModule
  exec: contract.UtilsModule
  exec: contract.VaultModule
  exec: contract.InitialCoreProxy
  exec: router.AccountRouter
  exec: router.CoreRouter
  exec: router.USDRouter
  exec: invoke.upgrade_core_proxy
  exec: provision.oracle_manager
Checking IPFS for package oracle-manager:3.2.1-dev.f440e7fa...
    exec: contract.NodeModule
    exec: contract.InitialProxy
    exec: router.OracleRouter
    exec: invoke.upgrade_proxy
  exec: invoke.enable_basic_features
  exec: invoke.init_account
  exec: invoke.init_usd
  exec: invoke.register_const_one_oracle
  exec: invoke.set_oracle_manager
  exec: invoke.configure_usd_collateral
exec: invoke.allowAllCreatePool
exec: invoke.allowAllRegisterMarket
exec: invoke.createAccount
exec: invoke.registerCollateralPriceNode
exec: invoke.setMinimumLiquidityRatio
exec: provision.spotMarket
Checking IPFS for package synthetix-spot-market:latest...
  exec: import.synthetix
Checking IPFS for package synthetix:latest...
  exec: contract.AsyncOrderConfigurationModule
  exec: contract.AsyncOrderModule
  exec: contract.AsyncOrderSettlementModule
  exec: contract.AtomicOrderModule
  exec: contract.MarketConfigurationModule
  exec: contract.SpotMarketFactoryModule
  exec: contract.WrapperModule
  exec: contract.InitialSpotMarketProxy
  exec: router.SpotMarketRouter
  exec: router.SynthRouter
  exec: invoke.upgradeSpotMarketProxy
  exec: invoke.addCreateSynthToFeatureFlag
  exec: invoke.setSynthImplementation
  exec: invoke.setSynthetixSystem
  exec: invoke.addSpotMarketToFeatureFlag
exec: invoke.configureTknCollateral
exec: invoke.createPool
exec: invoke.createSynth
exec: invoke.mintCollateral
exec: provision.perpsMarket
Checking IPFS for package synthetix-perps-market:latest...
  exec: import.spotMarket
Checking IPFS for package synthetix-spot-market:latest...
  exec: contract.AsyncOrderModule
  exec: contract.AsyncOrderSettlementModule
  exec: contract.CoreModule
  exec: contract.FeatureFlagModule
  exec: contract.GlobalPerpsMarketModule
  exec: contract.LiquidationModule
  exec: contract.MarketConfigurationModule
  exec: contract.PerpsAccountModule
  exec: contract.PerpsMarketFactoryModule
  exec: contract.PerpsMarketModule
  exec: import.synthetix
Checking IPFS for package synthetix:latest...
  exec: contract.InitialProxy
  exec: router.PerpsMarketRouter
  exec: invoke.upgrade_proxy
  exec: invoke.addPerpsMarketToFeatureFlag
  exec: invoke.init_account
  exec: invoke.setSpotMarket
  exec: invoke.setSynthetix
  exec: invoke.addCreateMarketToFeatureFlag
exec: invoke.addAtomicFixedFee
exec: invoke.allowAllCreatePerpsMarket
exec: invoke.allowCreatePerpsAccount
exec: invoke.approveDeposit
exec: invoke.setMarketCollateralSupplyCap
exec: invoke.setMaxCollateralAmount
exec: invoke.setPreferredPool
exec: invoke.updateSynthPriceData
exec: invoke.createPerpsMarket
exec: invoke.depositCollateral
exec: invoke.initializeWrapper
exec: invoke.setPoolConfiguration
exec: invoke.delegateCollateral
exec: invoke.updatePriceFeed
exec: invoke.mintStablecoins

TRANSACTIONS:
╔══════════════════════════════╤════════════════════════════════════════════════════════════════════╗
║ createSynth                  │ 0xfca7673c34a2fd00bbd42201e1fd820dd9b1993f7a7374040b1492a27822a227 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ addAtomicFixedFee            │ 0xe916b2348a7cd84c4ca1d48a122bc55c9c0f3dd472460ff91452122f0b80adb6 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowAllCreatePerpsMarket    │ 0x2853f03700b95eb81852c15b57185f4c7ac87d9dafb29cf959b679da49b7a7a6 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowAllRegisterMarket       │ 0x498994487073749c1699da7f28f687a0bdb2382bbed5f38e711cbf846d984125 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ registerCollateralPriceNode  │ 0x8813a4e399110888ffe9b3cfd3e99b104dc4a8038471809f44c8e453dcd6f946 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ configureTknCollateral       │ 0xc95609edbab07ac7b79042d932c56dc946b7948f2121464824a4e94139a76dfc ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setMarketCollateralSupplyCap │ 0xd2ee436ae3f2f679ed1f3c798aa1792442f88bfa91250b1999c9188c9fdfe654 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ initializeWrapper            │ 0x3afe4e28d43a19d6ded3278cd29e47b51530172ebd71297fd933d0e457233da3 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ createAccount                │ 0x93b03111e547311073e2423f4d5793ba687ff7923b65ed2b6e5d024c9f22c8ce ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ mintCollateral               │ 0x7e4a2a9089c99a454c2bf7602c186abf1dacf97753e25d5b275cd4b9244ac130 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ approveDeposit               │ 0x8f4227e9d47514c0b74935dee2430d1f2725a1c267e7a4b9665cc42e26022c59 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ depositCollateral            │ 0x7b75af7c2257a1cbbbcd9b85dffb862d34c2b7731ef3a68541008c63309e0643 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ delegateCollateral           │ 0x2fdb4b039c65fa5e0bb6a0068edabdbb2bcc86e95c23d8f47f03da7c4942f31f ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ mintStablecoins              │ 0x6aba9c256b7c52b46f747ccafc0acd1e38ceb094872f3d522614d8ed9b1da667 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setMaxCollateralAmount       │ 0x6c315f955f94b187832a6824b29c6fa78792c551fb9667fc1fa7b1f5236994aa ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setMinimumLiquidityRatio     │ 0xdfb4cf8ed0ccb6037a07dbb21fa709faf6c88bac496eb555ceb200ed70cd5caa ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowAllCreatePool           │ 0x65197791e11b89bd07fa8b00428772a0fd60c1ec1156d0049c6856af526fc11f ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ createPool                   │ 0xfe023123e4c569af8c64377099e31ad407ae60e19442e3ba3c1c5292bee786ca ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ updateSynthPriceData         │ 0x199682473c9bbaa1306c701e4cbcf8398c368487a08040919e77d25078090a0e ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setPoolConfiguration         │ 0xcfc75dfc52aadcea94b2991a13dc2cd2dd1702227ddce9031796bc7f370dff5e ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ setPreferredPool             │ 0x77d8d835614f6b864c3cc8fe83fb7d9f6f3b822c2ebfc725714b90498cf73add ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ allowCreatePerpsAccount      │ 0x7ae1afbcde8f46c3953a65ccf772cb3a6d2a0ed28c5762cbb620ed41b4ca1e60 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ createPerpsMarket            │ 0xd44ef0e649f20eb8ea5050a454b4fec9b47bb04f17415cd10b3f2c0a8d478a22 ║
╟──────────────────────────────┼────────────────────────────────────────────────────────────────────╢
║ updatePriceFeed              │ 0x73b6706868f85fcfbe2c7ac81382a00671477e830e022f59dd2739fbffea2484 ║
╚══════════════════════════════╧════════════════════════════════════════════════════════════════════╝


EXTRA DATA:
╔═══════════════╤════════════════════════════════════════════════════════════════════╗
║ synthMarketId │ 1                                                                  ║
╟───────────────┼────────────────────────────────────────────────────────────────────╢
║ tkn_oracle_id │ 0x6cb1042857d7329f5ee7ae88c96fd693a39b006b6a8884139c99802c856ca588 ║
╟───────────────┼────────────────────────────────────────────────────────────────────╢
║ perpsMarketId │ 2                                                                  ║
╚═══════════════╧════════════════════════════════════════════════════════════════════╝

Successfully built package synthetix-sandbox:3.1.2 (ipfs://QmcZRdePJHgT5Xo2V1HWj8P9aCfWJWyoALJncCEzSFk8K5)

```

File: packages/builder/src/actions.ts [hunk 7]
```diff
@@ -1,14 +1,15 @@
-import Ajv, { JTDDataType, SomeJTDSchemaType } from 'ajv/dist/jtd';
 import { ChainBuilderRuntime } from './runtime';
 
-const ajv = new Ajv();
-
 import contractSpec from './steps/contract';
+
```
Comment (dbeal-eth) [inline packages/builder/src/actions.ts:4]: random extra spaces being added here?

File: packages/builder/src/actions.ts [hunk 8]
```diff
@@ -82,11 +58,15 @@ export function registerAction(action: CannonAction) {
...
 
-export function getChainDefinitionValidator() {
-  return ajv.compile(ChainDefinitionSchema);
+  (
+    chainDefinitionSchema.pick({
+      description: true,
+      keywords: true,
+      setting: true,
+      import: true,
+    }) as any
```
Comment (dbeal-eth) [inline packages/builder/src/actions.ts:68]: what is the purpose of using `.pick` and then `as any` afterwards here?

File: packages/builder/src/definition.ts [hunk 9]
```diff
@@ -25,15 +25,18 @@ export type StateLayers = {
...
+/**
+ * Cannonfile metadata layout
+ */
 export class ChainDefinition {
   private raw: RawChainDefinition;
 
   readonly allActionNames: string[];
 
-  // actions which have no dependencies
+  /** actions which have no dependencies */
```
Comment (dbeal-eth) [inline packages/builder/src/definition.ts:36]: is jekyll not compatible with `//` and you have to use `/** */`?

File: packages/builder/src/schemas.zod.ts [hunk 10]
```diff
@@ -0,0 +1,304 @@
+import { z } from 'zod';
+import { handleZodErrors } from './error/zod';
+
+/// ================================ INPUT CONFIG SCHEMAS ================================ \\\
+
+export const contractSchema = z
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:6]: so for the builder base cannon steps its all well to put the zod definitions in this file. However, what happens for custom steps for validation (ex. `cannon-plugin-router` https://github.com/Synthetixio/cannon-plugin-router )

I *personally* am not a big fan of putting all the types for a package into a single file like some sort of a dumping ground

File: packages/builder/src/schemas.zod.ts [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+  .object({
+    artifact: z.string(),
+  })
+  .merge(
+    z
+      .object({
+        create2: z.boolean(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:13]: can we share an example of how we would add documentation description text to these types? I would like to do so myself if we think its out of scope, but it would be great to be able to just encode descriptions for what all of these fields do right in the code!

File: packages/builder/src/schemas.zod.ts [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        from: z.string(),
+        nonce: z.string(),
+        abi: z.string(),
+        abiOf: z.array(z.string()),
+        args: z.array(z.any()),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:18]: might actually be possible to constrain this type a little bit going forward. for example the type could be like `type argsType = string | number | argsType[] | Record<string, argsType>`

Comment (dbeal-eth) [reply packages/builder/src/schemas.zod.ts:18]: i know this is not how the previous ajv definition worked but looking at minor improvements

File: packages/builder/src/schemas.zod.ts [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+
+export const contractSchema = z
+  .object({
+    artifact: z.string(),
+  })
+  .merge(
+    z
+      .object({
+        create2: z.boolean(),
+        from: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:14]: could we have a zod custom validation for an address? the validation would just be `ethers.utils.isAddress(f)`. Definitely something that can be verified as part of syntax

File: packages/builder/src/schemas.zod.ts [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+        nonce: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:15]: even though the type is a string, it also must be numeric (either 0x prefixed hex or regular decimal number)

File: packages/builder/src/schemas.zod.ts [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+        abi: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:16]: should be JSON parsable even though it is a string

File: packages/builder/src/schemas.zod.ts [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+        abiOf: z.array(z.string()),
+        args: z.array(z.any()),
+        libraries: z.record(z.string()),
+
+        // used to force new copy of a contract (not actually used)
+        salt: z.string(),
+
+        value: z.string(),
+        overrides: z.object({
+          gasLimit: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:26]: same as nonce this has to be numeric

File: packages/builder/src/schemas.zod.ts [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+        }),
+
+        depends: z.array(z.string()),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:29]: this shuold be validated to be `<step type>.<name>` format. can probably do regex match here at least.

File: packages/builder/src/schemas.zod.ts [hunk 18]
```diff
@@ -0,0 +1,304 @@
...
+        chainId: z.number().int(),
+        preset: z.string(),
+        depends: z.array(z.string()),
+      })
+      .deepPartial()
+  );
+
+export const invokeSchema = z
+  .object({
+    target: z.array(z.string()).nonempty(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:50]: validate that its either an address or valid cannon step name (cannon step names can only have certain characters I believe?)

File: packages/builder/src/schemas.zod.ts [hunk 19]
```diff
@@ -0,0 +1,304 @@
...
+    func: z.string(),
+  })
+  .merge(
+    z
+      .object({
+        abi: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:56]: shoudl be json parsable

File: packages/builder/src/schemas.zod.ts [hunk 20]
```diff
@@ -0,0 +1,304 @@
...
+
+        args: z.array(z.any()),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:58]: coudl be a bit more constrained than any

File: packages/builder/src/schemas.zod.ts [hunk 21]
```diff
@@ -0,0 +1,304 @@
...
+        from: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:59]: must be address

File: packages/builder/src/schemas.zod.ts [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+        fromCall: z
+          .object({
+            func: z.string(),
+          })
+          .merge(
+            z.object({
+              args: z.array(z.any()),
+            })
+          ),
+        value: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:69]: must be numeric. could be decimal (ie `1.234`)

File: packages/builder/src/schemas.zod.ts (hunk 21 above)
Comment (dbeal-eth) [reply packages/builder/src/schemas.zod.ts:59]: could also be interpolation

File: packages/builder/src/schemas.zod.ts [hunk 23]
```diff
@@ -0,0 +1,304 @@
...
+  .merge(
+    z
+      .object({
+        abi: z.string(),
+
+        args: z.array(z.any()),
+        from: z.string(),
+        fromCall: z
+          .object({
+            func: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:62]: the usage of `.deepPartial` on line 94 will cause `func` to be optional here if I am not mistaken, but the field is not optional if `fromCall` is specified

Comment (dbeal-eth) [reply packages/builder/src/schemas.zod.ts:62]: https://github.com/colinhacks/zod#deeppartial

File: packages/builder/src/schemas.zod.ts [hunk 24]
```diff
@@ -0,0 +1,304 @@
...
+          })
+        ),
+        depends: z.array(z.string()),
+      })
+      .deepPartial()
+  );
+
+export const provisionSchema = z
+  .object({
+    source: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:99]: validate this has cannon package format (or interpolation) `packageName:packageVersion`

File: packages/builder/src/schemas.zod.ts [hunk 25]
```diff
@@ -0,0 +1,304 @@
...
+      })
+      .deepPartial()
+  );
+
+/**
+ * @internal NOTE: if you edit this schema, please also edit the constructor of ChainDefinition in 'definition.ts' to account for non-action components
+ */
+export const chainDefinitionSchema = z
+  .object({
+    name: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:132]: name must not be greater than 31 characters and must have no special characters https://github.com/usecannon/cannon/blob/main/packages/registry/contracts/CannonRegistry.sol#L31

File: packages/builder/src/schemas.zod.ts [hunk 26]
```diff
@@ -0,0 +1,304 @@
...
+    version: z.string(),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:133]: version must not be longer than 31 characters and may have other restrictions

File: packages/builder/src/schemas.zod.ts [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+  .merge(
+    z
+      .object({
+        description: z.string(),
+        keywords: z.array(z.string()),
+        setting: z.record(
+          z
+            .object({
+              description: z.string(),
+              type: z.enum(['number', 'string', 'boolean']),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:144]: type probably isnt needed at this point, we havent ever used it tmk. do you think it still has a use?

File: packages/builder/src/schemas.zod.ts [hunk 28]
```diff
@@ -0,0 +1,304 @@
...
+              defaultValue: z.string(),
+            })
+            .partial()
+        ),
+        import: z.object({ importSchema }),
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:149]: why is import schema specially included here?

File: packages/builder/src/schemas.zod.ts [hunk 29]
```diff
@@ -0,0 +1,304 @@
...
+      })
+      .deepPartial()
+  );
+
+export const runSchema = z
```
Comment (dbeal-eth) [inline packages/builder/src/schemas.zod.ts:154]: should be removed, the run schema is only in cli

File: packages/builder/src/steps/contract.ts [hunk 30]
```diff
@@ -108,6 +82,8 @@ export default {
   },
 
   configInject(ctx: ChainBuilderContextWithHelpers, config: Config) {
+    validateConfig('contract', config);
```
Comment (dbeal-eth) [inline packages/builder/src/steps/contract.ts:85]: rather than verifying the config at the beginning here why not change the `ChainDefinition` class to automatically validate before calling config inject? otherwise, this is likely to be missed by plugin integrators.

File: packages/builder/src/steps/invoke.ts [hunk 31]
```diff
@@ -245,10 +193,12 @@ export default {
   },
 
   configInject(ctx: ChainBuilderContextWithHelpers, config: Config) {
+    validateConfig('invoke', config);
+
     config = _.cloneDeep(config);
 
     if (config.target) {
-      config.target = config.target.map((v) => _.template(v)(ctx));
+      config.target = config.target.map((v) => _.template(v)(ctx)) as [string, ...string[]];
```
Comment (dbeal-eth) [inline packages/builder/src/steps/invoke.ts:201]: comment explaining the use of unusual type would be helpful here. is this supposed to be a `string[]` with min length of 1?

**Review by dbeal-eth — CHANGES_REQUESTED** [review]
almost all the `.string()` defined types in the file can be further validated in some way, either as an intterpolated value, or as the ultimate type they should be. see comments.

do we have CI to verify that the generated doc pages are up to date?

im suer this was an intentional design decision, but please include the reasoning for having the jekyll docs checked into the repo (especially when github is not hosting I assume?)

btw I know we talked earlier but doing the detailed review I realized somethings. most of my comments are *suggestions* and/oir not serious.

**Comment by FuzzB0t** [comment]
> Contributor

Good point, noted. Will add further validation to the string types.

As for the CI for docs I added this after our call today, I changed a bit about how the docs structure works and added that as part of it. Will push it along with the rest of the requested changes.

Thanks for your input! I think all of these suggestions are reasonable. The detailed review is super helpful.

> almost all the `.string()` defined types in the file can be further validated in some way, either as an intterpolated value, or as the ultimate type they should be. see comments.
> 
> do we have CI to verify that the generated doc pages are up to date?
> 
> im suer this was an intentional design decision, but please include the reasoning for having the jekyll docs checked into the repo (especially when github is not hosting I assume?)
> 
> btw I know we talked earlier but doing the detailed review I realized somethings. most of my comments are _suggestions_ and/oir not serious.

File: packages/builder/src/schemas.zod.ts (hunk 28 above)
Comment (FuzzB0t) [reply packages/builder/src/schemas.zod.ts:149]: This is meant to be the importSpec returned by the import action

File: packages/builder/src/schemas.zod.ts (hunk 11 above)
Comment (FuzzB0t) [reply packages/builder/src/schemas.zod.ts:13]: Yes will add this into my next batch of changes.

File: packages/builder/src/schemas.zod.ts (hunk 10 above)
Comment (FuzzB0t) [reply packages/builder/src/schemas.zod.ts:6]: Hmmm thats a good point. Existing plugins will have to switch their schemas to zod. However its worth rethinking a few things based on this:

- Before migrating, we validated any incoming "action" config schema as a JSON schema with properties and optional properties, any parameter was valid as long as it followed basic JTDDataType guidelines.          
- I realized that the ConfigValidationSchema that is used to validate any action schema the way I wrote it is probably not gonna work, since we need to sort of be able to do a "catchall" on config properties created for custom plugins.

My suggestion for that bug is to change the ConfigValidationSchema to be way less strict and see if there's a way for us to allow for new parameters in custom actions.

As for putting everything in one file: I don't personally like it either, it just makes generating the docs a bit easier but i can achieve the same output with a little extra work if we prefer to have the types in their respective action file.

File: packages/builder/src/schemas.zod.ts (hunk 27 above)
Comment (FuzzB0t) [reply packages/builder/src/schemas.zod.ts:144]: Havent really seen it used anywhere, so it's probably ok to remove. Will double check to make sure there isnt any package that happens to be using it so we avoid introducing breaking changes.

File: packages/builder/src/steps/invoke.ts (hunk 31 above)
Comment (FuzzB0t) [reply packages/builder/src/steps/invoke.ts:201]: Will do. It means exactly what you described. This is the type that zod returns when setting that validation.

File: packages/builder/src/definition.ts (hunk 9 above)
Comment (FuzzB0t) [reply packages/builder/src/definition.ts:36]: No this is to enable TSDOC comments which help with documenting certain things. Also have some useful features like being able to see these comments on instances of this class.

File: packages/builder/src/actions.ts (hunk 7 above)
Comment (FuzzB0t) [reply packages/builder/src/actions.ts:4]: Yep good catch, just removed these.

File: packages/builder/src/actions.ts (hunk 8 above)
Comment (FuzzB0t) [reply packages/builder/src/actions.ts:68]: Was trying to match the way it was being done before, also, as any was the only way I could index the object returned by `pick` with the `label` string

File: packages/builder/src/schemas.zod.ts (hunk 11 above)
Comment (FuzzB0t) [reply packages/builder/src/schemas.zod.ts:13]: Here's a guide on how to do this with typedoc: 

https://typedoc.org/guides/doccomments/

File: packages/builder/README.md
Comment (FuzzB0t) [inline packages/builder/README.md:1]: Misc change, will fix this.

File: packages/builder/src/schemas.zod.ts (hunk 18 above)
Comment (FuzzB0t) [reply packages/builder/src/schemas.zod.ts:105]: I can validate that its either an address or interpolated value, if we start to validate "known" step names it gets tricky since new plugins can add new step names

**Review by dbeal-eth — APPROVED** [review]

**Comment by FuzzB0t** [comment]
An important thing to note about merging this PR:

The cannon-plugin-router's schemas will have to be changed to zod schemas for it to work on this version.

**Comment by FuzzB0t** [comment]
This PR is ready to merge, could use one more review but I've gone pretty far testing manually. Will create a new issue to add a test suite to test for more edge cases and refine validations even further.
